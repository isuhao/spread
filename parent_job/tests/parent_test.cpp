#include "common.cpp"

struct TestJob : ParentJob
{
  int i;

  TestJob(int _i) : i(_i) { cout << "Making i=" << i << endl; }
  ~TestJob() { cout << "Killing i=" << i << endl; }

  void doJob()
  {
    if(i==3) setError("Job3");
    if(i==4) throw runtime_error("Job4");
    if(i==20)
      {
        while(!checkStatus()) Thread::sleep(0.01);
      }
    if(i==10)
      {
        setBusy("Job10");
        jobs.insert(JobPtr(new TestJob(11)));
        jobs.insert(JobPtr(new TestJob(12)));
        done.push_back(JobPtr(new TestJob(1)));
        done.push_back(JobPtr(new TestJob(2)));
        done.push_back(JobPtr(new TestJob(3)));
        done.push_back(JobPtr(new TestJob(4)));

        done[0]->getInfo()->abort();
        Thread::run(done[0]);
        Thread::run(done[1]);
        Thread::run(done[2]);
        Thread::run(done[3]);
        Thread::sleep(0.3);

        // This job is auto-aborted when we exit
        JobPtr p(new TestJob(20));
        Thread::run(p);
        jobs.insert(p);
      }
    if(i==30)
      {
        while(!checkStatus())
          {
            cout << "Thread-unlocked\n";
            Thread::sleep(0.02);
            boost::lock_guard<boost::mutex> lock(mutex);
            cout << "Thread-locked\n";
            Thread::sleep(0.02);
          }
      }
    if(checkStatus()) return;
    setDone();
  }
};

int main()
{
  cout << "Hello\n";

  TestJob *t = new TestJob(10);
  JobPtr ptr(t);
  print(ptr);
  t->run();
  print(ptr);
  ptr.reset();
  t = new TestJob(30);
  ptr.reset(t);
  Thread::run(ptr);
  Thread::sleep(0.3);
  {
    boost::lock_guard<boost::mutex> lock(t->mutex);
    cout << "LOCKED!\n";
    Thread::sleep(0.5);
  }
  cout << "UNLOCKED\n";
  Thread::sleep(0.3);
  cout << "Aborting...\n";
  ptr->getInfo()->abort();
  cout << "Waiting...\n";
  ptr->getInfo()->wait();
  cout << "EXIT\n";

  return 0;
}
