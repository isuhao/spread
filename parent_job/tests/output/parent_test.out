Hello
Making i=10
UNK:  0/0
Making i=11
Making i=12
Making i=1
Making i=2
Making i=3
Making i=4
Making i=20
SUC: Success 0/0
  DONE:
  ABO:  0/0
  SUC: Success 0/0
  ERR: Job3 0/0
  ERR: Job4 0/0
  JOBS:
  UNK:  0/0
  UNK:  0/0
  ABO:  0/0
Killing i=10
Killing i=1
Killing i=2
Killing i=3
Killing i=4
Killing i=20
Killing i=12
Killing i=11
Making i=30
Thread-unlocked
Thread-locked
Thread-unlocked
Thread-locked
Thread-unlocked
Thread-locked
Thread-unlocked
Thread-locked
Thread-unlocked
Thread-locked
Thread-unlocked
Thread-locked
Thread-unlocked
Thread-locked
Thread-unlocked
LOCKED!
UNLOCKED
Thread-locked
Thread-unlocked
Thread-locked
Thread-unlocked
Thread-locked
Thread-unlocked
Thread-locked
Thread-unlocked
Thread-locked
Thread-unlocked
Thread-locked
Thread-unlocked
Thread-locked
Thread-unlocked
Thread-locked
Aborting...
Waiting...
EXIT
Killing i=30
